build.rs

```diff
@@ -42,36 +42,151 @@ fn main() {
+fn build_dylint_driver(
+    _manifest_dir: PathBuf,
+    _out_dir: PathBuf,
+    dylint_driver_dst_file: PathBuf,
+) -> Result<()> {
+    // For `clippy` runs it is not necessary to build the `dylint` driver.
+    // Furthermore the fixed Rust nightly specified in `ink_linting/rust-toolchain`
+    // contains a bug that results in an `error[E0786]: found invalid metadata files` ICE.
+    //
+    // We still have to create an empty file though, due to the `include_bytes!` macro.
+    File::create(dylint_driver_dst_file)
+        .map_err(|err| {
+            anyhow::anyhow!(
+                "Failed creating an empty ink-dylint-driver.zip file: {:?}",
+                err
+            )
+        })
+        .map(|_| ())
+}
+
+/// Builds the crate in `ink_linting/`. This crate contains the `dylint` driver with ink! specific
+/// linting rules.
+#[cfg(not(feature = "cargo-clippy"))]
+fn build_dylint_driver(
+    manifest_dir: PathBuf,
+    out_dir: PathBuf,
+    dylint_driver_dst_file: PathBuf,
+) -> Result<()> {
+    let ink_dylint_driver_dir = manifest_dir.join("ink_linting");
+
+    let mut cmd = Command::new("cargo");
+
+    let manifest_arg = format!(
+        "--manifest-path={}",
+        ink_dylint_driver_dir.join("Cargo.toml").display()
+    );
+    let target_dir = format!("--target-dir={}", out_dir.display());
+    cmd.args(vec!["build", "--release", &target_dir, &manifest_arg]);
+
+    // We need to remove those environment variables because `dylint` uses a
+    // fixed Rust toolchain via the `ink_linting/rust-toolchain` file. By removing
+    // these env variables we avoid issues with different Rust toolchains
+    // interfering with each other.
+    cmd.env_remove("RUSTUP_TOOLCHAIN");
+    cmd.env_remove("CARGO_TARGET_DIR");
+
+    println!(
+        "Setting cargo working dir to '{}'",
+        ink_dylint_driver_dir.display()
     );
+    cmd.current_dir(ink_dylint_driver_dir.clone());
+
+    println!("Invoking cargo: {:?}", cmd);
+
+    let child = cmd
+        // capture the stdout to return from this function as bytes
+        .stdout(std::process::Stdio::piped())
+        .spawn()?;
+    let output = child.wait_with_output()?;
+
+    if !output.status.success() {
+        anyhow::bail!(
+            "`{:?}` failed with exit code: {:?}",
+            cmd,
+            output.status.code()
+        );
+    }
+
+    println!(
+        "Creating ink-dylint-driver.zip: destination archive '{}'",
+        dylint_driver_dst_file.display()
+    );
+
+    match zip_dylint_driver(
+        &out_dir.join("release"),
+        &dylint_driver_dst_file,
+        CompressionMethod::Stored,
+    ) {
+        Ok(_) => {
+            println!(
+                "done: {} written to {}",
+                ink_dylint_driver_dir.display(),
+                dylint_driver_dst_file.display()
+            );
+            Ok(())
+        }
+        Err(e) => {
+            eprintln!("Error: {:?}", e);
+            Err(e)
+        }
+    }
```

Comment:
To avoid destructuring the result here, could add another high level method below main which does roughly:

```
fn zip_template_and_build_dylint_driver() -> Result<()> {
  zip_template()?;
  build_dylint_driver()?;
  println!("Some success message which used to be in the Ok arm");
  Ok(())
}
```
Then at the top level in `main()` you can do the `eprintln!` which will handle all errors, and only need to destructure once. The `println!`s in the `Ok` arm can just move inline to the methods.